File: _Enutrof_[__-1]🎧🌝💉☕/Day 49/word-search.py
class Solution:
    def search(self, r, c, i):
        if i == len(self.word):
            return True
        if (
            min(r, c) < 0
            or r >= self.row_length
            or c >= self.col_length
            or self.word[i] != self.board[r][c]
            or (r, c) in self.path
        ):
            return False
        self.path.add((r, c))
        res = (
            self.search(r + 1, c, i + 1)
            or self.search(r - 1, c, i + 1)
            or self.search(r, c + 1, i + 1)
            or self.search(r, c - 1, i + 1)
        )
        self.path.remove((r, c))
        return res
    def exist(self, board: List[List[str]], word: str) -> bool:
        self.word = word
        self.board = board
        self.row_length, self.col_length = len(board), len(board[0])
        self.path = set()

        count = defaultdict(int, sum(map(Counter, board), Counter()))
        if count[word[0]] > count[word[-1]]:
            word = word[::-1]
            
        for r in range(self.row_length):
            for c in range(self.col_length):
                if self.search(r, c, 0):
                    return True
        return False

File: olix/day-49/word-search.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function (board, word) {
  let m = board.length;
  let n = board[0].length;
  let visited = new Array(m);
  for (let i = 0; i < m; i++) {
    visited[i] = new Array(n).fill(false);
  }
  let dfs = function (i, j, k) {
    if (i < 0 || i >= m || j < 0 || j >= n || board[i][j] !== word[k]) {
      return false;
    }
    if (k === word.length - 1) {
      return true;
    }
    visited[i][j] = true;
    let res =
      dfs(i + 1, j, k + 1) ||
      dfs(i - 1, j, k + 1) ||
      dfs(i, j + 1, k + 1) ||
      dfs(i, j - 1, k + 1);
    visited[i][j] = false;
    return res;
  };
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (dfs(i, j, 0)) {
        return true;
      }
    }
  }
  return false;
};

console.log(
  exist(
    [
      ["A", "B", "C", "E"],
      ["S", "F", "C", "S"],
      ["A", "D", "E", "E"],
    ],
    "ABCCED"
  )
);
